packages/integration-aws-node/src/DefaultAwsCredentialsProvider.ts

```diff
@@ -0,0 +1,274 @@
+  }
+
+  if (config.profile) {
+    return getProfileCredentials(config.profile!, config.region);
+  }
+
+  return getDefaultCredentialsChain();
+}
+
+/**
+ * Constructs the credential provider needed by the AWS SDK for the main account
+ *
+ * Order of precedence:
+ * 1. Static creds
+ * 2. Profile creds
+ * 3. Default AWS SDK creds chain
+ */
+function getMainAccountCredentialsProvider(
+  config: AwsIntegrationMainAccountConfig,
+): CredentialProvider {
+  if (config.accessKeyId) {
+    return getStaticCredentials(config.accessKeyId!, config.secretAccessKey!);
+  }
+
+  if (config.profile) {
+    return getProfileCredentials(config.profile!, config.region);
+  }
+
+  return getDefaultCredentialsChain();
+}
+
+/**
+ * Handles the creation and caching of credential providers for AWS accounts.
+ *
+ * @public
+ */
+export class DefaultAwsCredentialsProvider implements AwsCredentialsProvider {
+  static fromConfig(config: Config): DefaultAwsCredentialsProvider {
+    const awsConfig = config.has('aws')
+      ? readAwsIntegrationConfig(config.getConfig('aws'))
+      : {
+          accounts: [],
+          mainAccount: {},
+          accountDefaults: {},
+        };
+
+    const mainAccountProvider = getMainAccountCredentialsProvider(
+      awsConfig.mainAccount,
+    );
+    const mainAccountCreds: AwsCredentials = {
+      provider: mainAccountProvider,
+    };
+
+    const accountCreds = new Map<string, AwsCredentials>();
+    for (const accountConfig of awsConfig.accounts) {
+      const provider = getAccountCredentialsProvider(
+        accountConfig,
+        mainAccountCreds.provider,
+      );
+      accountCreds.set(accountConfig.accountId, {
+        accountId: accountConfig.accountId,
+        stsRegion: accountConfig.region,
+        provider,
+      });
+    }
+
+    return new DefaultAwsCredentialsProvider(
+      accountCreds,
+      awsConfig.accountDefaults,
+      mainAccountCreds,
+    );
+  }
+
+  private constructor(
+    private readonly accountCredentials: Map<string, AwsCredentials>,
+    private readonly accountDefaults: AwsIntegrationDefaultAccountConfig,
+    private readonly mainAccountCredentials: AwsCredentials,
+  ) {}
+
+  /**
+   * Returns {@link AwsCredentials} for a given AWS account.
+   *
+   * @example
+   * ```ts
+   * const { provider } = await getCredentials({
+   *   accountId: '0123456789012',
+   * })
+   *
+   * const { provider } = await getCredentials({
+   *   arn: 'arn:aws:ecs:us-west-2:123456789012:service/my-http-service'
+   * })
+   * ```
+   *
+   * @param opts - the AWS account ID or AWS resource ARN
+   * @returns A promise of {@link AwsCredentials}.
+   */
+  async getCredentials(
+    opts: AwsCredentialsProviderOptions,
+  ): Promise<AwsCredentials> {
+    // Determine the account ID: either explicitly provided or extracted from the provided ARN
+    let accountId = opts.accountId;
+    if (opts.arn && !accountId) {
+      const arnComponents = parse(opts.arn);
```

Comment:
I think it might be useful to expose a ARN helper from this package, as this is most certainly a behaviour most AWS integrations need (i.e. for splitting resource names out of an ARN etc.). It might be nicer, to expose a helper via this package, instead of having to import `import { parse } from '@aws-sdk/util-arn-parser';` in every package?

Reply:
Considering your pseudo-code example here: https://github.com/backstage/backstage/pull/14125#discussion_r1020609260 a common usecase not covered by the package today is the parsing of resource names: `const {region, resourceName} = parse(resourceArn);
   const tableName = resourceName.split("/")[1];`

Reply:
True, but that logic is not really specific to Backstage, and that specific logic isn't universal across AWS services.  It depends on the specific AWS service the plugin is integrating with.  Unfortunately there is a handful of different flavors of ARN formats that change the logic for how the resource name is extracted.

The AWS CDK has some logic around extracting resource name from various ARN formats.  We could copy that, but ideally this would go into the SDK's util-arn-parser library, instead of copying it into a Backstage-specific library:
https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/core/lib/arn.ts#L6-L48

Reply:
I have opened an issue in the aws sdk repo for this:
https://github.com/aws/aws-sdk-js-v3/issues/4243